### PR TITLE
#4: Implement basic reddit auth using passport.js

### PR DESCRIPTION
**Pull request by xrisk**
Functions for:

1. Validating the reddit user against the DB
2. Serializing the user object into the session
3. Handling new user signup from reddit

have not been implemented yet and have been marked with TODO.

### CODE REVIEW

File: packages/backend/package.json [hunk 1]
```diff
@@ -7,7 +7,6 @@
   "scripts": {
     "test": "jest",
     "build": "babel src -d build",
-    "dev:server": "nodemon src/index.js --exec babel-node",
```
Comment (asn007) [inline packages/backend/package.json]: Any reason for removal?

File: packages/backend/src/server.js [hunk 2]
```diff
@@ -67,6 +79,17 @@ export function createServer(bind) {
       res.json({ data: { healthy: true } })
     );
 
+    app.get('/auth_status', (req, res) =>
```
Comment (asn007) [inline packages/backend/src/server.js]: We have API for it :p

**Review by asn007 — CHANGES_REQUESTED** [review]
Fix my comments and we're good I think
P. S. Maybe erase session.state on callback too

File: packages/backend/src/server.js (hunk 2 above)
Comment (xrisk) [reply packages/backend/src/server.js]: Oh I wrote this as kind of a utility endpoint...

Comment (asn007) [reply packages/backend/src/server.js]: Move it into api/auth and it's good

File: packages/backend/src/api/1/handlers/auth.js [hunk 3]
```diff
@@ -1,3 +1,33 @@
+import crypto from 'crypto';
 import express from 'express';
+import passport from 'passport';
 
-module.exports = express.Router();
+const router = express.Router();
+
+router.get('/reddit/callback', (req, res, next) => {
+  // Check for origin via state token
+  if (req.query.state === req.session.state)
```
Comment (asn007) [inline packages/backend/src/api/1/handlers/auth.js]: Actually from what I see, you can extract it into a middleware chain for that request and the code will get a little more readable imo

File: packages/backend/src/api/1/handlers/auth.js [hunk 4]
```diff
@@ -1,3 +1,33 @@
...
+router.get('/reddit', (req, res, next) => {
+  if (!req.session.state)
+    req.session.state = crypto.randomBytes(32).toString('hex');
+  passport.authenticate('reddit', {
+    state: req.session.state,
+    duration: 'permanent',
+  })(req, res, next);
+});
+
+router.get('/auth_status', (req, res) =>
```
Comment (asn007) [inline packages/backend/src/api/1/handlers/auth.js]: Maybe just /status?

File: packages/backend/src/api/1/handlers/auth.js [hunk 5]
```diff
@@ -1,3 +1,33 @@
...
+  res.json({
+    auth_status: req.isAuthenticated(),
```
Comment (asn007) [inline packages/backend/src/api/1/handlers/auth.js]: You can actually just check if user is null / not null, so I doubt auth_status field is really needed

File: packages/backend/src/helpers/setupPassportStrategies.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+const RedditStrategy = require('passport-reddit').Strategy;
+
+const setupPassportStrategies = passport => {
+  passport.use(
+    new RedditStrategy(
+      {
+        clientID: 'rWjd7fOjL7eq3w',
```
Comment (asn007) [inline packages/backend/src/helpers/setupPassportStrategies.js:7]: Are these id / secret host bound? 
If so, we can actually keep it as part of our devenv

File: packages/backend/src/server.js [hunk 7]
```diff
@@ -67,6 +79,11 @@ export function createServer(bind) {
       res.json({ data: { healthy: true } })
     );
 
+    app.get('/login_failure', (req, res) => {
```
Comment (asn007) [inline packages/backend/src/server.js]: Can be moved to API as well I think

**Review by asn007 — CHANGES_REQUESTED** [review]

File: packages/backend/src/helpers/setupPassportStrategies.js (hunk 6 above)
Comment (xrisk) [reply packages/backend/src/helpers/setupPassportStrategies.js:7]: They are bound to my reddit account and the callback has to be specified in Reddit, but Reddit allows modifying the callback URL.

File: packages/backend/src/api/1/handlers/auth.js [hunk 8]
```diff
@@ -1,3 +1,48 @@
+import crypto from 'crypto';
 import express from 'express';
+import passport from 'passport';
 
-module.exports = express.Router();
+const router = express.Router();
+
+router.get(
```
Comment (asn007) [inline packages/backend/src/api/1/handlers/auth.js:7]: Ain't really what I was asking.
I'll explain in PR comments more concise, but:
separate the endpoints so they're
/reddit
and
/reddit/callback

**Review by asn007 — APPROVED** [review]

**Review by asn007 — CHANGES_REQUESTED** [review]
Will explain what I mean in comments to PR

**Comment by asn007** [comment]
You could actually abstract away that if() { passport(req, res, next) } ugly chain by utilising middleware in the following way:
```js
router.get('/reddit/callback', (req, res, next) => next(req.session.state !== req.query.state ? new Error(403) : undefined), passport.authenticate('reddit', { someData }))
```

**Comment by asn007** [comment]
Approved

[Approved with PullApprove](https://pullapprove.com/rTeenDeveloper/hub/pull-request/23/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=rTeenDeveloper/hub)

**Pull request closed without merging**